ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/manager/snap/SnapServer.java

```diff
@@ -14,66 +14,616 @@
+      final ProtocolContext protocolContext) {
+
+    this.snapServerEnabled = snapConfig.isSnapServerEnabled();
     this.snapMessages = snapMessages;
-    this.worldStateArchive = worldStateArchive;
-    this.registerResponseConstructors();
+    this.worldStateStorageCoordinator = worldStateStorageCoordinator;
+    this.protocolContext = Optional.of(protocolContext);
+    registerResponseConstructors();
+
+    // subscribe to initial sync completed events to start/stop snap server:
+    this.protocolContext
+        .flatMap(ProtocolContext::getSynchronizer)
+        .filter(z -> z instanceof DefaultSynchronizer)
+        .map(DefaultSynchronizer.class::cast)
+        .ifPresentOrElse(
+            z -> this.listenerId.set(z.subscribeInitialSync(this)),
+            () -> LOGGER.warn("SnapServer created without reference to sync status"));
+  }
+
+  /**
+   * Create a snap server without registering a listener for worldstate initial sync events or
+   * priming worldstates by root hash. Used by unit tests.
+   */
+  @VisibleForTesting
+  SnapServer(
+      final EthMessages snapMessages,
+      final WorldStateStorageCoordinator worldStateStorageCoordinator,
+      final Function<Hash, Optional<BonsaiWorldStateKeyValueStorage>> worldStateStorageProvider) {
+    this.snapServerEnabled = true;
+    this.snapMessages = snapMessages;
+    this.worldStateStorageCoordinator = worldStateStorageCoordinator;
+    this.worldStateStorageProvider = worldStateStorageProvider;
+    this.protocolContext = Optional.empty();
+  }
+
+  @Override
+  public void onInitialSyncCompleted() {
+    start();
+  }
+
+  @Override
+  public void onInitialSyncRestart() {
+    stop();
+  }
+
+  public synchronized SnapServer start() {
+    if (!isStarted.get() && snapServerEnabled) {
+      // if we are bonsai and full flat, we can provide a worldstate storage:
+      var worldStateKeyValueStorage = worldStateStorageCoordinator.worldStateKeyValueStorage();
+      if (worldStateKeyValueStorage.getDataStorageFormat().equals(DataStorageFormat.BONSAI)
+          && worldStateStorageCoordinator.isMatchingFlatMode(FlatDbMode.FULL)) {
+        LOGGER.debug("Starting snap server with Bonsai full flat db");
+        var bonsaiArchive =
+            protocolContext
+                .map(ProtocolContext::getWorldStateArchive)
+                .map(BonsaiWorldStateProvider.class::cast);
+        var cachedStorageManagerOpt =
+            bonsaiArchive.map(archive -> archive.getCachedWorldStorageManager());
+
+        if (cachedStorageManagerOpt.isPresent()) {
+          var cachedStorageManager = cachedStorageManagerOpt.get();
+          this.worldStateStorageProvider = cachedStorageManager::getStorageByRootHash;
+
+          // when we start we need to build the cache of latest 128 worldstates
+          // trielogs-to-root-hash:
+          var blockchain = protocolContext.map(ProtocolContext::getBlockchain).orElse(null);
```

Comment:
why we are not also preloading the bonsai worldstate for this headers ?

Reply:
we can, and I had it that way for a while to pass hive tests.  But actually loading the last 128 worldstates delays startup a bit and feels a little clunky.  Just priming the blockhash to roothash cache seems to be a happy medium